### ISSUE TITLE
[$300] When a user registers, set bookmark by default

### ISSUE THREAD

**Issue by jmgasper** [topcoder-platform/challenge-forum-processor#17]
@atelomycterus - One thing we'd like to investigate with the forum processor (this repo) is the ability to set bookmarks on both `Code Documents` and `Code Questions` when members register.  This way the registered users will automatically get email notifications of new posts.

This is similar to the current forum functionality, where a user has "watches" set on those forums when they register for a challenge - we want to do the same in Vanilla, if possible.

**Comment by jmgasper** [comment]
Contest https://www.topcoder.com/challenges/30122437 has been created for this ticket.<br/><br/>```This is an automated message for ghostar via Topcoder X```

**Comment by jmgasper** [comment]
Contest https://www.topcoder.com/challenges/30122437 has been updated - it has been assigned to obog.<br/><br/>```This is an automated message for ghostar via Topcoder X```

**Comment by atelomycterus** [comment]
@jmgasper 

> 'This is similar to the current forum functionality, where a user has "watches" set on those forums when they register for a challenge - we want to do the same in Vanilla, if possible.'

This functionality is named "follow the category" in Vanilla. 
If User follows individual categories and be notified of all posts within them.
Vanilla REST API has '/categories/{id}/follow' (PUT) but it doesn't support userId and works the current user only.

Swagger API: http://vanilla.topcoder-dev.com/settings/swagger

$this->getSession()->UserID is hard coded in CategortiesAPIController.php
		
So the solution:
- Create a TopcoderAPIController for APIv2 Endpoints. The class should extends AbstractApiControllers.
The controller should be in https://github.com/topcoder-platform/forums/vanilla/applications/vanilla/controllers/api
We don't copy vanilla sources, just follow the structure.
- Add topcoder.yml in 'openapi'.  https://github.com/topcoder-platform/forums/vanilla/applications/vanilla/openapi
- Copy contents of  the `vanilla` folder in dockerfile

Then we can use it in challenge-forum-processor.

> 'set bookmarks on both Code Documents and Code Questions when members register'

Vanilla uses 'a bookmark' for discussions only. It also works the current user only.

**Comment by jmgasper** [comment]
Ok, I'm alright with that.  Will bump the payment here for the extra work.

**Comment by atelomycterus** [comment]
@jmgasper Keep you updated on it. Couple of issues with Vanilla notifications.

The 'UserMeta' table is used to notify users. It keeps userId ,  categoryId  and notificationType.  
Challenge forum processor adds a record in 'UserMeta'. 
![image](https://user-images.githubusercontent.com/10196632/79753915-5e725f00-831f-11ea-8477-aeb9bafa3c9d.png)

After post/comment added, Vanilla fetches users from the 'UserMeta' table by categoryId to notify them.  But it doesn't use the target category. It finds the parent category with depth <=2 and uses it instead if the target category.
So User is not notified because the depth of `Code documents`\`Code Questions` is 4. 
If we add userId + `Development Forums` (the depth is 2) in the UserMeta  
then this will not work  as  the current forum functionality.
User will be notified when people posts/comments on other challenge categories.

The main issues that the depth value is hard coded in several places (class.disscusionmodel.php, class.commentmodel.php). We can see only 2 level categories on the Preferences page because it's also hard coded in class.hooks.php:
![image](https://user-images.githubusercontent.com/10196632/79753977-7a760080-831f-11ea-93a0-522b25726d9c.png)

So I'm trying to fix it with minor changes or without any changes in Vanilla.

**Comment by jmgasper** [comment]
@atelomycterus - Sounds good, thanks.  I bumped the payment again due to the additional work.

**Comment by atelomycterus** [comment]
@jmgasper Thanks a lot.  Implemented. Plan to commit in the next hour.  Couple questions:
1. Do we need to send any notification if a category or its parent categories archived?

2. > This way the registered users will automatically get email notifications of new posts.

 Should User be notified by email only? We can add popup notification as well. 


**Developer Notes:**
1) Don't use 'Vanilla.QueueNotifications'. It's removed from core. https://github.com/vanilla/vanilla/pull/9522 
2) That's not possible to override  methods of Models (DiscussionModel and so on).
It's possible to add new methods only:
"Magic methods only work in classes that extend Gdn_Pluggable". The DiscussionModel extends Gdn_Model which in turn extends Gdn_Pluggable.
"Magic methods allow you to create new methods and add them to existing objects."
Other classes (formatters, rules, middlewares) can be injected into a container.

3) The current version of Vanilla doesn't support sending batch emails. They might include it the next release.  

4) Upgrade Vanilla to the next release:
Model classes were fixed to support notifications for any categories. 
Need to update the model classes from Vanilla installation because these classes are the core of Vanilla forum:
`/vanilla/applications/vanilla/model/class.discussionmodel.php`
`/vanilla/applications/vanilla/model/class.commentmodel.php.` 
Check if our fixes are required. 

5) the Vanilla endpoints `/categories/{id}/follow` only change flags in the 'UserCategory' table. It's not used in Vanilla. There are no any notifications/actions.
 
**Emails**

Configure email notifications in Admin Dashboard:
1) Check email settings
![image](https://user-images.githubusercontent.com/10196632/79848306-96ce7780-83c9-11ea-903b-76da15e8c170.png)

2) Send a test email
![image](https://user-images.githubusercontent.com/10196632/79848412-b9609080-83c9-11ea-971b-58f5122377f7.png)

3) Default email template can be configured:
![image](https://user-images.githubusercontent.com/10196632/79848489-d006e780-83c9-11ea-8329-027551861a16.png)


_Developer notes:_
Preferences are saved in the 'UserMeta' table. This table is updated by challenge forum processor.

Posts/comments are added by user => if it's required any notification then a new record will be added in the Activity table. The email notification status is saved in  the 'Emailed' column. The status:
```
    /** Activity status: The activity was added before this system was put in place. */
    const SENT_ARCHIVE = 1;
    /** Activity status: The activity sent just fine. */
    const SENT_OK = 2;
    /** Activity status: The activity is waiting to be sent. */
    const SENT_PENDING = 3;
    /** Activity status: The activity could not be sent. */
    const SENT_FAIL = 4;
    /** Activity status: There was an error sending the activity, but it can be retried. */
    const SENT_ERROR = 5;
    /** Activity status: The recipient was not eligible for an email notification. */
    const SENT_SKIPPED = 6;
    /** Activity status: Sending is in progress. */
    const SENT_INPROGRESS = 31;
```

**Comment by atelomycterus** [comment]
@jmgasper  just for your information: the Activity table where all notifications saved is pruned after 2 months by default. It can be configured but a large number of records will cause performance issues.

**Comment by jmgasper** [comment]
Ok, sounds good.  Is the merge request ready to go?

**Comment by atelomycterus** [comment]
@jmgasper yes, I was waiting for confirmation. Need to apply 2 PRs:
https://github.com/topcoder-platform/challenge-forum-processor/pull/21
https://github.com/topcoder-platform/forums/pull/14
Thanks!

**Comment by jmgasper** [comment]
@atelomycterus - Thanks.  I've deployed this, and I'm testing it, but when I send a message like this in Kafka:

```
{
  "topic": "challenge.notification.events",
  "originator": "topcoder-challenges-api",
  "timestamp": "2020-04-22T00:22:41.164Z",
  "mime-type": "application/json",
  "payload": {
    "type": "USER_REGISTRATION",
    "data": {
      "challengeId": "2893c84b-282c-4405-a70f-cdfd3730b82d",
      "userId": 151743
    }
  }
}
```

I am seeing this error in the challenge-forum-processor logs.  Is my message bad?  Did you get this to work?

**Comment by jmgasper** [comment]
I'm wondering if this is a config issue - I'm checking that as well.

**Comment by atelomycterus** [comment]
@jmgasper I tested with challenge.action.resource.create.json/challenge.action.resource.delete.json.

> I am seeing this error in the challenge-forum-processor logs. Is my message bad? Did you get this to work?
> 
> 2020-04-23T06:12:09.048651+00:00 app[worker.1]: 2020-04-23T06:12:09.048Z [debug]: Message from challenge.notification.events: [{"type":"USER_REGISTRATION","data":{"challengeId":"2893c84b-282c-4405-a70f-cdfd3730b82d","userId":151743}}]
> 2020-04-23T06:12:09.204068+00:00 app[worker.1]: 2020-04-23T06:12:09.203Z [error]: Error: Unknown Error
> 2020-04-23T06:12:09.204071+00:00 app[worker.1]: at Request._callback (/app/node_modules/tc-core-library-js/lib/auth/m2m.js:70:24)
> 2020-04-23T06:12:09.204071+00:00 app[worker.1]: at Request.self.callback (/app/node_modules/request/request.js:185:22)
> 2020-04-23T06:12:09.204072+00:00 app[worker.1]: at Request.emit (events.js:310:20)
> 2020-04-23T06:12:09.204073+00:00 app[worker.1]: at Request.<anonymous> (/app/node_modules/request/request.js:1154:10)
> 2020-04-23T06:12:09.204073+00:00 app[worker.1]: at Request.emit (events.js:310:20)
> 2020-04-23T06:12:09.204074+00:00 app[worker.1]: at IncomingMessage.<anonymous> (/app/node_modules/request/request.js:1076:12)
> 2020-04-23T06:12:09.204074+00:00 app[worker.1]: at Object.onceWrapper (events.js:416:28)
> 2020-04-23T06:12:09.204075+00:00 app[worker.1]: at IncomingMessage.emit (events.js:322:22)
> 2020-04-23T06:12:09.204076+00:00 app[worker.1]: at endReadableNT (_stream_readable.js:1187:12)
> 2020-04-23T06:12:09.204076+00:00 app[worker.1]: at processTicksAndRejections (internal/process/task_queues.js:84:21)

I got the same error. From node_modules/tc-core-library-js/lib/auth/m2m.js
```
       //Check the token expiry  
        if (appCachedToken) {
          if (getTokenExpiryTime(appCachedToken) <= 0) {
            appCachedTokenExpired = true
          }
        }
        if (!appCachedToken || appCachedTokenExpired ) {
            request.post(options, function (error, response, body) {
              if (error) {
                return reject(new Error(error))
              }
              if (body.access_token) {
                cachedToken[clientId] = body.access_token
                resolve(cachedToken[clientId])
              }
              else {
                reject(new Error('Unknown Error'))
              }
            })
        }
        else {
          resolve(appCachedToken)
        }
```

**Comment by jmgasper** [comment]
@atelomycterus - I narrowed this down to a bad config.  I'm working with the Topcoder team to get new creds for Auth0.

**Comment by jmgasper** [comment]
@atelomycterus - It may be a few days / next week before we can fully test this.  I'm going to close this out now so that you get paid properly, and if we run into any issues while testing, after we fix things up in the config, I'll open up new ticket(s) for you.

**Comment by jmgasper** [comment]
Payment failed: [500]: Failed to get billing detail for the project.<br/><br/>```This is an automated message for ghostar via Topcoder X```

**Comment by atelomycterus** [comment]
@jmgasper Could you check this payment for 'When a user registers, set bookmark by default'? The fix was accepted 8 days ago. Let me know if you get new creds for Auth0, i can test it.

**Comment by jmgasper** [comment]
Payment task has been updated: https://software.topcoder.com/review/actions/ViewProjectDetails?pid=30122437<br/><br/>```This is an automated message for ghostar via Topcoder X```
